library/Icinga/Web/RememberMe.php

```diff
@@ -0,0 +1,318 @@
+     *
+     * @param string $username
+     * @param string $password
+     *
+     * @return static
+     */
+    public static function fromCredentials($username, $password)
+    {
+        $aesCrypt = new AesCrypt();
+        $rememberMe = new static();
+        $rememberMe->encryptedPassword = $aesCrypt->encryptToBase64($password);
+        $rememberMe->username = $username;
+        $rememberMe->aesCrypt = $aesCrypt;
+
+        return $rememberMe;
+    }
+
+    /**
+     * Remove expired remember me information from the database
+     */
+    public static function removeExpired()
+    {
+        (new static())->getDb()->delete(static::TABLE, [
+            'expires_at < NOW()'
+        ]);
+    }
+
+    /**
+     * Get the remember me cookie
+     *
+     * @return Cookie
+     */
+    public function getCookie()
+    {
+        return (new Cookie(static::COOKIE))
+            ->setExpire($this->getExpiresAt())
+            ->setHttpOnly(true)
+            ->setValue(implode('|', [
+                $this->encryptedPassword,
+                base64_encode($this->aesCrypt->getTag()),
+                base64_encode($this->aesCrypt->getIV()),
+            ]));
+    }
+
+    /**
+     * Get the timestamp when the cookie expires
+     *
+     * Defaults to now plus 30 days, if not set via {@link setExpiresAt()}.
+     *
+     * @return int
+     */
+    public function getExpiresAt()
+    {
+        if ($this->expiresAt === null) {
+            $this->expiresAt = time() + 60 * 60 * 24 * 30;
+        }
+
+        return $this->expiresAt;
+    }
+
+    /**
+     * Set the timestamp when the cookie expires
+     *
+     * @param int $expiresAt
+     *
+     * @return $this
+     */
+    public function setExpiresAt($expiresAt)
+    {
+        $this->expiresAt = $expiresAt;
+
+        return $this;
+    }
+
+    /**
+     * Authenticate via the remember me cookie
+     *
+     * @return bool
+     *
+     * @throws \Icinga\Exception\AuthenticationException
+     */
+
+    public function authenticate()
+    {
+        $password = $this->aesCrypt->decryptFromBase64($this->encryptedPassword);
+        $auth = Auth::getInstance();
+        $authChain = $auth->getAuthChain();
+        $authChain->setSkipExternalBackends(true);
+        $user = new User($this->username);
+        if (! $user->hasDomain()) {
+            $user->setDomain(Config::app()->get('authentication', 'default_domain'));
+        }
+        $authenticated = $authChain->authenticate($user, $password);
+        if ($authenticated) {
+            $auth->setAuthenticated($user);
+        }
+
+        return $authenticated;
+    }
+
+    /**
+     * Persist the remember me information into the database
+     *
+     * Any previous stored information is automatically removed.
+     *
+     * @param null $iv
+     *
+     * @return $this
+     */
+    public function persist($iv = null)
+    {
+        $this->remove($iv);
```

Comment:
what happens if `$iv` is null? If you provide `null` as default, you have to guarantee the method call succeeds. Currently it doesn't, or runs a `DELETE` for no reason.

Reply:
If $iv is null, nothing will be deleted. Should I rather make an if($iv) condition?